docs/design/keyboard_handler.md

```diff
@@ -0,0 +1,87 @@
+# Keyboard handler
+Package providing ability to handle keyboard input via simple interface with callbacks. 
+
+## Goal
+We need to be able to handle keyboard input in unified way with cross-platform implementation.
+
+## Design Proposal
+It would be great to have some simple interface to be able to subscribe to the specific key 
+press event using callbacks. Also it would useful to have ability to handle events for 
+combination of keys pressed in one callback as well as have possibility to subscribe to the 
+same event from multiple clients.
+
+The following pseudocode suggests the high level API for aforementioned design proposal.
+```cpp
+    ClientClass1 client1;
+    ClientClass2 client2;
+    KeyboardHandler keyboard_handler;
+    client1.register_key_press_callbacks(keyboard_handler);
+    client2.register_callbacks_for_keyboard_handler(keyboard_handler);
+```
+Inside client class registering for the callbacks could be organized as simple as:
+```cpp
+  ClientClass1::register_key_press_callbacks(KeyboardHandler & keyboard_handler) 
+  {
+    keyboard_handler.add_key_press_callback(callback_fn, KeyboardHandler::KeyCode::CURSOR_UP);
+    keyboard_handler.add_key_press_callback(callback_fn, KeyboardHandler::KeyCode::SHIFT_F1);
```

Comment:
Maybe this is "key + optional modifier", instead of defining predetermined combinations of keys:

```
# Subscribe to Shift+Ctrl+F1
keyboard_handler.subscribe(callback=callback_fn, key=KeyCode::F1, modifiers=KeyMod::SHIFT | KeyMod::CTRL);
```

Reply:
It would great to have such possibility to use such key modifiers which could be applied with "or" mask. But it is almost impossible or very difficult to implement in a cross-platform approach. \
Also in this case will be an open question how to detect to which key and modifier corresponds event when callback will arrive?
```cpp
 void callback_function(KeyboardHandler::KeyCode key_code)
  {
    using KeyCode = KeyboardHandler::KeyCode;
    switch (key_code) {
    // ...
```

Reply:
Ok I've tried to consider adding key modifiers for representing `SHIFT`, `ALT` and `CTRL`. \
It is possible to add this feature, although in this case it will add constraint on the values in `enum class KeyboardHandlerBase::KeyCode` that key codes with modifiers should be strictly ordered and follow in sequence in the same order as original key codes without modifiers. It will be much more error prone if someone will try to add new key codes to the enum and it's impossible to check if key codes will be out of order or if one them will be missed.

Giving that I have a strong opinion against such initiative. And I think that explicit prefix e.g. `SHIFT_F1` looks not so bad and pretty clear for usage.

Reply:
> n this case it will add constraint on the values in enum class KeyboardHandlerBase::KeyCode that key codes with modifiers should be strictly ordered and follow in sequence in the same order as original key codes without modifier

I was confused by this - I'm not quite sure what you mean.

The way I am thinking about it, you have something like:

```
enum KeyCode:
  a,
  b,
  c,
  F1,
  LEFT_BRACKET,
  etc...

enum KeyMod:
  CTRL,
  SHIFT,
  ALT
```

This is very easy to define - there is a linear number of each type, and events are given by combination. This is how I understand that most input libraries handle this interface.

The method you're proposing, e.g. `SHIFT_F1` means that you have to define M*M*M*N enum entries, where M is number of keymods and N is number of keycodes. This is not scalable to an entire keyboard.

Reply:
I'm looking at https://github.com/torvalds/linux/blob/master/include/uapi/linux/input-event-codes.h for example of keycodes.

For pseudocode:

```
func on_keydown_event(KeyCode key) {
  if (key in MODIFIER_KEYS) {
    current_mods_ |= key;
  } else {
    auto callback = registered_callbacks_.find(current_mods_, key);
    if (callback) {
      callback();    
    }
  }
}

func on_keyup_event(KeyCode key) {
  if (key in MODIFIER_KEYS) {
    current_mods_ &= ~key;
  }
}


```

Reply:
It easy to define in design but not so simple to implement.

The problem is that I don't have such a low level API to be able to detect if some key modifier got pressed.
```cpp
if (key in MODIFIER_KEYS) {
    current_mods_ |= key;
```
Look on this
https://github.com/MichaelOrlov/rosbag2/blob/michaelorlov/keyboard_handler/keyboard_handler/src/keyboard_handler_unix_impl.cpp#L104-L120

It's just returning the sequence of codes like array and for instance for 'a' key corresponds one byte = '97' for `CTRL`+'a' it also returns one byte with value = '1' . \
And there are no way to detect if just only key modifiers pressed. Read function doesn't respond on them. \
In case of functional keys it's even more sophisticated. For instance:
```
static constexpr char             F11[] = {27, 91, 50, 51, 126, '\0'};
static constexpr char SHIFT_F11[] = {27, 91, 50, 51, 59, 50, 126, '\0'};
```
`Shift` modifier adding `59, 50` somewhere in the middle. The situation is even worse keyboard handler running in terminal under some application. For instance CLION IDE modifying returning sequence for F11 (refer to the [note](https://github.com/MichaelOrlov/rosbag2/blob/michaelorlov/keyboard_handler/keyboard_handler/src/default_unix_key_map.cpp#L16-L22))
```
constexpr char F11[] = {27, 91, 50, 51, 126, 27, '\0'};  // Extra 27 at the end
```
In current implementation I have simple luck up table with hash table for detecting key codes. This is relatively simple and straight forward approach.  And it would be relatively easy to extend API to provide ability to alter LUT in runtime with custom key codes without recompilation source code in case if someone will need a quick workaround for redefining existing key binding or need to add new key code.

If go to the design approach when modifiers will be separate from physical keys it will require to write some parser for returning key sequences with some heuristic inside.  In this case if someone will need to change corresponding key mapping it will require to change logic in those parser.

I can try to implement some parser for key modifiers but I am not sure how it will align with cross platform specific. Give me a couple days to try it.

Reply:
@emersonknapp I am in the process of redesigning keyboard handler to be able support key modifiers separate from base key press.

I made implementation for Linux with basic alphanumeric keys, although supporting functional and control keys with key modifiers is non trivial and there are no descent solution for it. The difficulties that such modifiers as `SHIFT`, `CTRL` and `ALT` among inserting some extra symbols in the middle of sequence also changing base values for some keys.
https://github.com/MichaelOrlov/rosbag2/blob/michaelorlov/keyboard_handler/keyboard_handler/src/default_unix_key_map.cpp#L41-L67
For instance:
```cpp
static constexpr char F1[]  = {27, 79, 80, '\0'};
static constexpr char SHIFT_F1[]  = {27, 91, 49, 59, 50, 80, '\0'};
```
The second byte for `F1` changing from 79 to 91.

I tend to agree and accept idea for using new API with modifiers i.g.
```cpp
/// \brief Type for callback functions
  using callback_t = std::function<void (KeyCode, KeyModifiers)>;
```
but **without supporting key modifiers for functional and control keys**. \
I would propose to put this restriction as design/implementation limitation and known issue.

Are you ok with it?

Reply:
Yes, I am OK with that limitation - let's make sure to document it.

Reply:
@emersonknapp I've prepared new version of the keyboard handler with changed API. Please review it at https://github.com/ros-tooling/keyboard_handler/pull/1.

I've documented design/implementation limitations in Doxygen comments for relevant classes. Also going to move and update design document in a separate PR.

Reply:
@emersonknapp I've updated design document and created new PR https://github.com/ros-tooling/keyboard_handler/pull/2 in ros-tooling please review.

Reply:
Approved the keyboard_handler PR, let's close this one